### PR TITLE
Support cab archives that have paths as backslashes

### PR DESCRIPTION
**Pull request by hughsie**
This is a regression from when we switched to using GCab.

Refactor out the code into a new UploadedFile class that has unit tests, to
prevent this kind of thing happening again.

Fixes: https://github.com/hughsie/lvfs-website/issues/36

### CODE REVIEW

File: app/util.py [hunk 1]
```diff
@@ -4,17 +4,23 @@
...
+import os
 import calendar
 import datetime
 from glob import fnmatch
 
 from flask import session, request, flash, render_template
 
+def _get_basename_safe(fn):
+    """ gets the file basename, also with win32-style backslashes """
+    return os.path.basename(fn.replace('\\', '/'))
```
Comment (superm1) [inline app/util.py]: Would using `ntpath `work better for this?  Although I suppose archives with the slashes the other way would break in the same way.

Comment (hughsie) [reply app/util.py]: Right, python seems to really want to stick to the current OS type. I didn't find a better way to do this, and it seemed harmless enough.

**Comment by hughsie** [comment]
As an aside, do you know how to set up CI to run the tests for this trivial python case? I'm planning to add a lot more to cover the flask side of the app.

**Comment by superm1** [comment]
I think https://docs.travis-ci.com/user/languages/python/ should cover exactly what you need.  No needs to run in docker or anything.

Regarding when this fix goes into production, is it immediate or when you restart LVFS this weekend? (2018-02-04 at 20:00 GMT according to fwupd.org)

**Comment by hughsie** [comment]
I pushed it to the server just now as it was blocking you. At the weekend we should get a hardware upgrade all being well.

**Comment by hughsie** [comment]
> No needs to run in docker or anything.

How's that going to work with the GObjectIntrospection requirement?

**Comment by superm1** [comment]
Thanks.

I would think that you would just install the appropriate dependencies for gobject introspection during the "install" command.  If it does turn out to be too challenging on Ubuntu trusty (Travis CI env),  then a Docker container is obviously an option.

**Pull request merged**